### DISCUSSION TITLE
webpack support

### DISCUSSION THREAD

**Discussion by ayroblu** [pmndrs/jotai#2431] [Bug report]
## Summary

Hi, I noticed that `window.__NUMBER_OF_JOTAI_INSTANCES__` was injected into our console. From a look at the details, I'm guessing that this is a vite thing where they inject an environment variable. However we use webpack that doesn't have this automatically happen. Is there a way to handle the traditional webpack case (where it would typically use process.env.NODE_ENV === 'production')

## Link to reproduction

N/A

**Comment by dai-shi** [comment]
This is a one-size-does-not-fit-all situation.
I thought it would be nice for ESM build to be Node.js `process` free. Thus, it uses Vite convention.
Our CJS build still uses traditional `process.env.NODE_ENV` style.
We used to have `module` condition for webpack, which is dropped in v2. #1675
I wonder if webpack supports `webpack` condition or something.

Anyway, I will think about it and if anyone has some expertise, feel free to jump in.

**Reply by ayroblu** [reply]
I was thinking that the check here: https://github.com/pmndrs/jotai/blob/3ca1d435b543965bf1bcabe8487a2c7fc0e047a6/src/vanilla/store.ts#L848

changes from: 
`if (import.meta.env?.MODE !== 'production') {`
to
`if (import.meta.env && import.meta.env.MODE !== 'production') {`

So that the "default" is guess production if not vite rather than the reverse

**Reply by dai-shi** [reply]
Please see https://unpkg.com/browse/jotai@2.7.0/esm/vanilla.mjs for the transformed code.
<img width="873" alt="image" src="https://github.com/pmndrs/jotai/assets/490574/55a93864-f988-4eba-9f3c-8cb4442d73b1">

We used to make the default guess to be production, but with that many people missed warning messages and reported many issues, so we changed the default guess to be development.

**Reply by dai-shi** [reply]
By the way, one reason for the transformation is that our "src" code is to work with Deno.

**Comment by dai-shi** [comment]
I guess one workaround is to use cjs build with `require()`.
```js
const { atom } = require('jotai');
```
If you don't prefer the require syntax, I think you can configure webpack to use cjs build instead of esm.

Though, our cjs build is es5, so it might not be ideal.
When we move to Jotai v3, we probably drop es5, like https://github.com/pmndrs/zustand/pull/2380.

I wonder if there a way to configure webpack so that it replaces `import.meta.env`.

**Reply by ayroblu** [reply]
https://github.com/webpack/webpack/issues/15833
Not really yet, there's one example of a loader, but that's just rewriting the ast, which is reasonable but not an ideal final solution: https://github.com/XiaoLin1995/import-meta-loader/blob/main/src/transform.ts#L78-L80

I think there should be a way to support both syntaxes until this stuff becomes standardised

**Reply by dai-shi** [reply]
Can you try `require()` in the meantime and see if it works? (I'm also interested in how much the bundle size would increase.)

**Reply by ayroblu** [reply]
I think this isn't what we're gonna do, I do have a way in webpack to exclude it, but I already shipped the dev experience to prod, which is fine but a bit unfortunate. I just think that the default should be handled a bit better

**Comment by dai-shi** [comment]
Until https://github.com/webpack/webpack/issues/15833 lands,
#2434 is an attempt to provide a specific build for webpack.
It's with `webpack` export condition. It should be easier to configure (than using cjs build), even if it's not officially supported.

https://ci.codesandbox.io/status/pmndrs/jotai/pr/2434
☝️ See "Local Install Instructions"

**Reply by ayroblu** [reply]
```
yarn add v1.22.17
info No lockfile found.
(node:39700) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
[1/4] 🔍  Resolving packages...
error An unexpected error occurred: "https://pkg.csb.dev/pmndrs/jotai/commit/e4090cee/jotai: Request failed \"500 Internal Server Error\"".
info If you think this is a bug, please open a bug report with the information provided in "/Users/blu/ws/temp/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.
```
